File: server-java/src/main/java/bbs/Handlers.java
package bbs;

import chat.EnvelopeOuterClass.Envelope;
import org.zeromq.ZMQ;

import java.util.List;

public class Handlers {

    public static Envelope handleLogin(Envelope msg, Storage storage) {
        String username = msg.getUsername().trim();

        if (username.isBlank()) {
            return ProtocolUtil.makeMessage(
                    "LOGIN_REP", "", "", false,
                    "Nome de usuário inválido.",
                    null, msg.getRequestId(), "",
                    0, "", 0, null, 0
            );
        }

        storage.saveLogin(username, msg.getTimestamp());

        return ProtocolUtil.makeMessage(
                "LOGIN_REP", username, "", true,
                "", null, msg.getRequestId(), "",
                0, "", 0, null, 0
        );
    }

    public static Envelope handleListChannels(Envelope msg, Storage storage) {
        List<String> channels = storage.listChannels();

        return ProtocolUtil.makeMessage(
                "LIST_CHANNELS_REP", "", "", true,
                "", channels, msg.getRequestId(), "",
                0, "", 0, null, 0
        );
    }

    public static Envelope handleCreateChannel(
            Envelope msg,
            Storage storage,
            ZMQ.Socket pubSocket,
            LogicalClock logicalClock,
            String serverName,
            int serverRank,
            long physicalTime
    ) {
        String channelName = msg.getChannelName().trim();

        if (channelName.isBlank()) {
            return ProtocolUtil.makeMessage(
                    "CREATE_CHANNEL_REP", "", "", false,
                    "Nome do canal inválido.",
                    null, msg.getRequestId(), "",
                    0, "", 0, null, 0
            );
        }

        if (!channelName.matches("^[a-zA-Z0-9_-]+$")) {
            return ProtocolUtil.makeMessage(
                    "CREATE_CHANNEL_REP", "", "", false,
                    "Nome do canal deve conter apenas letras, números, '_' ou '-'.",
                    null, msg.getRequestId(), "",
                    0, "", 0, null, 0
            );
        }

        boolean created = storage.createChannel(channelName, msg.getTimestamp());

        if (!created) {
            return ProtocolUtil.makeMessage(
                    "CREATE_CHANNEL_REP", "", channelName, false,
                    "Canal já existe.",
                    null, msg.getRequestId(), "",
                    0, "", 0, null, 0
            );
        }

        replicateChannel(
                channelName,
                msg.getRequestId(),
                pubSocket,
                logicalClock,
                serverName,
                serverRank,
                physicalTime
        );

        return ProtocolUtil.makeMessage(
                "CREATE_CHANNEL_REP", "", channelName, true,
                "", null, msg.getRequestId(), "",
                0, "", 0, null, 0
        );
    }

    public static Envelope handlePublish(
            Envelope msg,
            Storage storage,
            ZMQ.Socket pubSocket,
            LogicalClock logicalClock,
            String serverName,
            int serverRank,
            long physicalTime
    ) {
        String username = msg.getUsername().trim();
        String channelName = msg.getChannelName().trim();
        String messageText = msg.getMessageText().trim();

        if (username.isBlank()) {
            return ProtocolUtil.makeMessage(
                    "PUBLISH_REP", "", "", false,
                    "Usuário inválido.",
                    null, msg.getRequestId(), "",
                    0, "", 0, null, 0
            );
        }

        if (channelName.isBlank()) {
            return ProtocolUtil.makeMessage(
                    "PUBLISH_REP", "", "", false,
                    "Canal inválido.",
                    null, msg.getRequestId(), "",
                    0, "", 0, null, 0
            );
        }

        /*
         * Como o broker usa round-robin, o cliente pode ter visto o canal em um
         * servidor, mas a publicação cair em outro servidor que ainda não recebeu
         * a réplica. Para manter o funcionamento distribuído, o servidor cria
         * localmente o canal antes de salvar a mensagem.
         */
        if (!storage.channelExists(channelName)) {
            storage.createChannel(channelName, msg.getTimestamp());
            replicateChannel(
                    channelName,
                    msg.getRequestId(),
                    pubSocket,
                    logicalClock,
                    serverName,
                    serverRank,
                    physicalTime
            );
        }

        if (messageText.isBlank()) {
            return ProtocolUtil.makeMessage(
                    "PUBLISH_REP", username, channelName, false,
                    "Mensagem inválida.",
                    null, msg.getRequestId(), "",
                    0, "", 0, null, 0
            );
        }

        storage.saveMessage(
                username,
                channelName,
                messageText,
                msg.getTimestamp(),
                msg.getRequestId()
        );

        logicalClock.tick();

        Envelope payload = ProtocolUtil.makeMessage(
                "CHANNEL_MESSAGE",
                username,
                channelName,
                true,
                "",
                null,
                null,
                messageText,
                logicalClock.getValue(),
                serverName,
                serverRank,
                null,
                physicalTime
        );

        pubSocket.sendMore(channelName);
        pubSocket.send(payload.toByteArray(), 0);

        replicateMessage(
                username,
                channelName,
                messageText,
                msg.getRequestId(),
                msg.getTimestamp(),
                pubSocket,
                logicalClock,
                serverName,
                serverRank,
                physicalTime
        );

        return ProtocolUtil.makeMessage(
                "PUBLISH_REP", username, channelName, true,
                "", null, msg.getRequestId(), messageText,
                0, "", 0, null, 0
        );
    }

    private static void replicateChannel(
            String channelName,
            String requestId,
            ZMQ.Socket pubSocket,
            LogicalClock logicalClock,
            String serverName,
            int serverRank,
            long physicalTime
    ) {
        logicalClock.tick();

        Envelope replication = ProtocolUtil.makeMessage(
                "REPLICATION_CHANNEL",
                "",
                channelName,
                true,
                "",
                null,
                requestId,
                "",
                logicalClock.getValue(),
                serverName,
                serverRank,
                null,
                physicalTime
        );

        pubSocket.sendMore("replication");
        pubSocket.send(replication.toByteArray(), 0);

        System.out.println(
                "Canal replicado no tópico 'replication' | canal="
                        + channelName
                        + " | origem="
                        + serverName
        );
    }

    private static void replicateMessage(
            String username,
            String channelName,
            String messageText,
            String requestId,
            long sentTimestamp,
            ZMQ.Socket pubSocket,
            LogicalClock logicalClock,
            String serverName,
            int serverRank,
            long physicalTime
    ) {
        logicalClock.tick();

        Envelope replication = ProtocolUtil.makeMessage(
                "REPLICATION_MESSAGE",
                username,
                channelName,
                true,
                "",
                null,
                requestId,
                messageText,
                logicalClock.getValue(),
                serverName,
                serverRank,
                null,
                physicalTime
        ).toBuilder()
                .setTimestamp(sentTimestamp)
                .build();

        pubSocket.sendMore("replication");
        pubSocket.send(replication.toByteArray(), 0);

        System.out.println(
                "Mensagem replicada no tópico 'replication' | id="
                        + requestId
                        + " | canal="
                        + channelName
                        + " | origem="
                        + serverName
        );
    }

    public static Envelope handleUnknown(Envelope msg) {
        return ProtocolUtil.makeMessage(
                "ERROR_REP", "", "", false,
                "Tipo de mensagem desconhecido: " + msg.getType(),
                null, msg.getRequestId(), "",
                0, "", 0, null, 0
        );
    }
}

File: server-python/app/handlers.py
import re

from protocol import make_message


CHANNEL_NAME_PATTERN = re.compile(r"^[a-zA-Z0-9_-]+$")


def handle_login(msg, storage):
    username = msg.username.strip()

    if not username:
        return make_message(
            msg_type="LOGIN_REP",
            success=False,
            error_message="Nome de usuário inválido.",
            request_id=msg.request_id,
        )

    storage.save_login(username=username, login_timestamp=msg.timestamp)

    return make_message(
        msg_type="LOGIN_REP",
        success=True,
        username=username,
        request_id=msg.request_id,
    )


def handle_list_channels(msg, storage):
    channels = storage.list_channels()

    return make_message(
        msg_type="LIST_CHANNELS_REP",
        success=True,
        channels=channels,
        request_id=msg.request_id,
    )


def handle_create_channel(
    msg,
    storage,
    pub_socket=None,
    logical_clock=None,
    server_name="",
    server_rank=0,
    physical_time=0,
):
    channel_name = msg.channel_name.strip()

    if not channel_name:
        return make_message(
            msg_type="CREATE_CHANNEL_REP",
            success=False,
            error_message="Nome do canal inválido.",
            request_id=msg.request_id,
        )

    if not CHANNEL_NAME_PATTERN.match(channel_name):
        return make_message(
            msg_type="CREATE_CHANNEL_REP",
            success=False,
            error_message="Nome do canal deve conter apenas letras, números, '_' ou '-'.",
            request_id=msg.request_id,
        )

    created = storage.create_channel(
        channel_name=channel_name,
        created_at=msg.timestamp,
    )

    if not created:
        return make_message(
            msg_type="CREATE_CHANNEL_REP",
            success=False,
            channel_name=channel_name,
            error_message="Canal já existe.",
            request_id=msg.request_id,
        )

    if pub_socket is not None and logical_clock is not None:
        replicate_channel(
            channel_name=channel_name,
            request_id=msg.request_id,
            pub_socket=pub_socket,
            logical_clock=logical_clock,
            server_name=server_name,
            server_rank=server_rank,
            physical_time=physical_time,
        )

    return make_message(
        msg_type="CREATE_CHANNEL_REP",
        success=True,
        channel_name=channel_name,
        request_id=msg.request_id,
    )


def handle_publish(
    msg,
    storage,
    pub_socket,
    logical_clock=None,
    server_name="",
    server_rank=0,
    physical_time=0,
):
    username = msg.username.strip()
    channel_name = msg.channel_name.strip()
    message_text = msg.message_text.strip()

    if not username:
        return make_message(
            msg_type="PUBLISH_REP",
            success=False,
            error_message="Usuário inválido.",
            request_id=msg.request_id,
        )

    if not channel_name:
        return make_message(
            msg_type="PUBLISH_REP",
            success=False,
            error_message="Canal inválido.",
            request_id=msg.request_id,
        )

    # Como o broker usa round-robin, a publicação pode cair em um servidor
    # que ainda não recebeu a réplica do canal. Para manter a consistência,
    # criamos o canal localmente antes de salvar a mensagem.
    if not storage.channel_exists(channel_name):
        storage.create_channel(
            channel_name=channel_name,
            created_at=msg.timestamp,
        )

        if logical_clock is not None:
            replicate_channel(
                channel_name=channel_name,
                request_id=msg.request_id,
                pub_socket=pub_socket,
                logical_clock=logical_clock,
                server_name=server_name,
                server_rank=server_rank,
                physical_time=physical_time,
            )

    if not message_text:
        return make_message(
            msg_type="PUBLISH_REP",
            success=False,
            error_message="Mensagem inválida.",
            request_id=msg.request_id,
        )

    storage.save_message(
        username=username,
        channel_name=channel_name,
        message_text=message_text,
        sent_timestamp=msg.timestamp,
        request_id=msg.request_id,
    )

    channel_payload = make_message(
        msg_type="CHANNEL_MESSAGE",
        username=username,
        channel_name=channel_name,
        message_text=message_text,
        logical_clock=msg.logical_clock,
        server_name=server_name,
        server_rank=server_rank,
        physical_time=physical_time,
    )

    pub_socket.send_multipart([
        channel_name.encode("utf-8"),
        channel_payload.SerializeToString(),
    ])

    if logical_clock is not None:
        replicate_message(
            username=username,
            channel_name=channel_name,
            message_text=message_text,
            request_id=msg.request_id,
            sent_timestamp=msg.timestamp,
            pub_socket=pub_socket,
            logical_clock=logical_clock,
            server_name=server_name,
            server_rank=server_rank,
            physical_time=physical_time,
        )

    return make_message(
        msg_type="PUBLISH_REP",
        success=True,
        username=username,
        channel_name=channel_name,
        message_text=message_text,
        request_id=msg.request_id,
    )


def replicate_channel(
    channel_name,
    request_id,
    pub_socket,
    logical_clock,
    server_name,
    server_rank,
    physical_time,
):
    logical_clock.tick()

    replication = make_message(
        msg_type="REPLICATION_CHANNEL",
        success=True,
        channel_name=channel_name,
        request_id=request_id,
        logical_clock=logical_clock.get_value(),
        server_name=server_name,
        server_rank=server_rank,
        physical_time=physical_time,
    )

    pub_socket.send_multipart([
        b"replication",
        replication.SerializeToString(),
    ])

    print(
        "Canal replicado no tópico 'replication' | "
        f"canal={channel_name} | origem={server_name}"
    )


def replicate_message(
    username,
    channel_name,
    message_text,
    request_id,
    sent_timestamp,
    pub_socket,
    logical_clock,
    server_name,
    server_rank,
    physical_time,
):
    logical_clock.tick()

    replication = make_message(
        msg_type="REPLICATION_MESSAGE",
        success=True,
        username=username,
        channel_name=channel_name,
        message_text=message_text,
        request_id=request_id,
        logical_clock=logical_clock.get_value(),
        server_name=server_name,
        server_rank=server_rank,
        physical_time=physical_time,
    )

    replication.timestamp = sent_timestamp

    pub_socket.send_multipart([
        b"replication",
        replication.SerializeToString(),
    ])

    print(
        "Mensagem replicada no tópico 'replication' | "
        f"id={request_id} | canal={channel_name} | origem={server_name}"
    )


def handle_unknown(msg):
    return make_message(
        msg_type="ERROR_REP",
        success=False,
        error_message=f"Tipo de mensagem desconhecido: {msg.type}",
        request_id=msg.request_id,
    )